juriscraper/pacer/iquery.py

```diff
@@ -0,0 +1,187 @@
+        #     <BR>
+        #   </CENTER>
+
+        # Rather than take the approach used by DocketParser of
+        # searching throughout the document for a docket number by
+        # regular expression, instead we take the <center> tag that follows
+        # the relevant cmecfMainContent <div>, and go line-by-line,
+        # delimited by <br>s. This approach is more structured:
+
+        center = self.tree.xpath('.//div[@id="cmecfMainContent"]//center')[0]
+        rows = self.redelimit_p(center, self.BR_REGEX)
+
+        # The first row demands special handling:
+        #   <B><FONT SIZE=+1>18-11572</FONT></B><B></B>Nancy Jean Stevens
+        # We take the docket number from the <font> tag (the innermost tag),
+        # although we could but have chosen the first <b> tag.
+        docket_number = force_unicode(rows[0].find('.//font').text_content())
+        # And case caption following the final <b></b> pair.
+        case_name_raw = force_unicode(rows[0].find('.//b[last()]').tail)
+
+        # Our job as a parser is to return the data, not to filter, clean,
+        # amend, or "harmonize" it. However downstream often expects harmonized
+        # data, so go with both.
+        case_name = clean_string(harmonize(case_name_raw))
+
+        # Iterate through the remaining rows, recognizing that the second row
+        # may be special (district court), in which case it lacks the
+        # <b> tag used for key/value pairs and it gives the judge's name.
+        # Such as:
+        #   Mark L. Wolf, presiding
+        #   <B>Date filed:</B> 02/10/2011
+        #   <B>Date filed:</B> 04/30/2018<B>Date of last filing:</B> 06/06/2018
+        data = {}
+        for i in xrange(1, len(rows)):
+            bolds = rows[i].findall('.//b')
+            if bolds is None:
+                if i == 1:
+                    # Second row, no bold => judge name!
+                    judge_role = force_unicode(rows[i].text_content())
+                    PRESIDING = ", presiding"
+                    assert judge_role.endswith(PRESIDING)
+                    data[u'judge_name'] = judge_role.rstrip(PRESIDING)
+                else:
+                    raise AssertionError('Line with no boldface?')
+            for bold in bolds:
+                boldtext = bold.text_content().strip()
+                assert boldtext.endswith(':')
+                field = boldtext.rstrip(':')
+                cleanfield = field.lower().replace(' ', '_').decode('utf-8')
+
+                value = bold.tail.strip()
+                data[cleanfield] = force_unicode(value)
+
+        # xxx use of dict.update() is kind of weird
+        data.update({
+            u'court_id': self.court_id,
+            u'docket_number': docket_number,
+            u'case_name': case_name,
+            u'case_name_raw': case_name_raw,
+        })
+
+        # I don't think this is a good idea, it's too indiscriminate
+        # data = clean_pacer_object(data)
```

Comment:
They might not be necessary in this instance, but these things all seem safe to me:

    1. Removing spaces before commas.
    1. Stripping whitespace from the ends.
    1. Normalizing white space.
    1. Forcing unicode.

?

Reply:
Pretty much none of those seem safe to me :)
Also, github doen't enumerate them for you in this contest, so they all show as `1.`.